File: app/src/main/java/com/example/networkaplication/details/DetailsRepositoryImpl.java
package com.example.networkaplication.details;

import android.app.Application;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Environment;

import com.example.networkaplication.models.select.MovieDetail;
import com.example.networkaplication.persistance.model.Details;
import com.example.networkaplication.persistance.repository.Repositories;
import com.example.networkaplication.persistance.repository.database.DatabaseRepositoryImpl;
import com.example.networkaplication.persistance.repository.database.dao.DetailsDao;
import com.example.networkaplication.retrofit.NetworkService;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class DetailsRepositoryImpl implements DetailsContract.DetailsRepository, Callback<MovieDetail>{

    private DetailsDao detailsDao;
    private MovieDetail movie;
    private DetailsCallback callback;

    private static File root = Environment.getExternalStorageDirectory();


    public DetailsRepositoryImpl(Application app, DetailsCallback callback) {
        DatabaseRepositoryImpl repository =
                (DatabaseRepositoryImpl) Repositories.getDatabase(app.getApplicationContext());

        this.callback = callback;

        detailsDao = repository.detailsDao();
    }

    @Override
    public void saveImage(Bitmap imageToSave, String omdbId) {
        File file = new File(root, omdbId.trim() + ".PNG");

        try {
            FileOutputStream out = new FileOutputStream(file);
            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
            imageToSave.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream);
            out.write(byteArrayOutputStream.toByteArray());

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public Uri loadImage(String omdbId) {
        File file = new File(root, omdbId.trim() + ".PNG");

        if (file.exists())
            return Uri.fromFile(file);

        return null;
    }

    @Override
    public void loadDetails(String filmName) {
        NetworkService.getInstance()
                .getOMDBApi()
                .getMovieDetails(filmName)
                .enqueue(this);
    }

    @Override
    public MovieDetail getMovie() {
        return this.movie;
    }

    @Override
    public Details getDetailFromId(String omdbId) {
        return detailsDao.findById(omdbId);
    }

    @Override
    public void onResponse(Call<MovieDetail> call, final Response<MovieDetail> response) {
        final MovieDetail movie = response.body();

        if (movie != null) {

            callback.onMoviesReceived(movie);
            String detailsLine = String
                    .format("%s %s %s %s/10 ", movie.getYear(), movie.getRuntime(),
                            movie.getCountry(), movie.getImdbRating());

            Details details = detailsDao.findByTitle(movie.getTitle());

            if (details == null) {
                Details details1 = new Details(movie.getImdbID());
                details1.setTitle(movie.getTitle());
                details1.setGenre(movie.getGenre());
                details1.setDetails(detailsLine);
                details1.setPlotSummary(movie.getPlot());
                callback.onDetailsReceived(details1);
                detailsDao.insert(details1);
            } else {
                details.setTitle(movie.getTitle());
                details.setDetails(detailsLine);
                details.setGenre(movie.getGenre());
                details.setPlotSummary(movie.getPlot());
                callback.onDetailsReceived(details);
                detailsDao.update(details);
            }
        }
    }

    @Override
    public void onFailure(Call<MovieDetail> call, Throwable t) {
        call.cancel();
    }
}

File: app/src/main/java/com/example/networkaplication/details/DetailsRepositoryImpl.kt
package com.example.networkaplication.details

import android.app.Application
import android.graphics.Bitmap
import android.net.Uri
import android.os.Environment
import com.example.networkaplication.App
import com.example.networkaplication.idling.EspressoIdlingResource
import com.example.networkaplication.models.select.MovieDetail
import com.example.networkaplication.persistance.model.Details
import com.example.networkaplication.persistance.repository.Repositories
import com.example.networkaplication.persistance.repository.database.DatabaseRepositoryImpl
import com.example.networkaplication.persistance.repository.database.dao.DetailsDao
import com.example.networkaplication.retrofit.OMDBApiInterface
import com.example.networkaplication.retrofit.component.DaggerDetailsComponent
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import javax.inject.Inject


class DetailsRepositoryImpl internal constructor(app: Application, private val callback: DetailsCallback) : DetailsContract.DetailsRepository, Callback<MovieDetail> {

    private val detailsDao: DetailsDao

    @Inject
    lateinit var omdbApiInterface: OMDBApiInterface


    init {
        val repository = Repositories.getDatabase(app.applicationContext) as DatabaseRepositoryImpl

        val component = DaggerDetailsComponent.builder()
                .appComponent((app as App).component)
                .build()
        component.inject(this)

        detailsDao = repository.detailsDao()
    }

    override fun saveImage(imageToSave: Bitmap, omdbId: String) {
        val file = File(root, omdbId.trim { it <= ' ' } + ".PNG")

        try {
            val out = FileOutputStream(file)
            val byteArrayOutputStream = ByteArrayOutputStream()
            imageToSave.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream)
            out.write(byteArrayOutputStream.toByteArray())

        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun loadImage(omdbId: String): Uri? {
        val file = File(root, omdbId.trim { it <= ' ' } + ".PNG")

        return if (file.exists()) Uri.fromFile(file) else null

    }

    override fun loadDetails(filmName: String) {
        EspressoIdlingResource.increment()

        omdbApiInterface.getMovieDetails(filmName).enqueue(this)
    }

    override fun getDetailFromId(omdbId: String): Details {
        return detailsDao.findById(omdbId)
    }

    override fun onResponse(call: Call<MovieDetail>, response: Response<MovieDetail>) {
        val movie = response.body()

        if (movie != null) {

            callback.onMoviesReceived(movie)
            val detailsLine = String
                    .format("%s %s %s %s/10 ", movie.year, movie.runtime,
                            movie.country, movie.imdbRating)

            val details = detailsDao.findByTitle(movie.title!!)

            if (details == null) {
                val details1 = Details(movie.imdbID!!)
                details1.title = movie.title
                details1.genre = movie.genre
                details1.details = detailsLine
                details1.plotSummary = movie.plot
//                callback.onDetailsReceived(details1)
                detailsDao.insert(details1)
            } else {
                details.title = movie.title
                details.genre = movie.genre
                details.details = detailsLine
                details.plotSummary = movie.plot
//                callback.onDetailsReceived(details)
                detailsDao.update(details)
            }

        }
    }

    override fun onFailure(call: Call<MovieDetail>, t: Throwable) {
        call.cancel()
    }

    companion object {

        private val root = Environment.getExternalStorageDirectory()
    }
}
